Added layout options

philphilphil opened the pull request:
Hello,
I added the ability to use layout options with task queries. Currently, it only supports "Hide Task Count" and "Hide Backlinks". I think it's a useful feature since for all but one Query I can hide both. This covers parts of #81.

I decided to add it to the query syntax via "layoutOptions:" and a comma separated list. I think like that it should be clear to the user, that it's not part of the filter.

```tasks
not done
path includes Arbeit
layoutOptions: hideTaskcount, hideBacklinks
```
The possibilities are endless and if more layout options will be added maybe it would make sense to refactor it into its own class to use within the query.

Let me know if this is something you would want in your project and/or if needs any changes.
Thanks!
Phil

Fixes #81

schemar commented:
OMG awesome, thank you. I will do a proper review when I have time.

Two notes:

1. Please add the new functionality to the README
2. Do you think the layoutOptions is in line with the other query options? 🤔I am not sure myself. The other ones „read“ like sentences, so maybe something like „hide task count“/„hide backlinks“ would be more consistent? What do you think?

In any case: thank you so much for this! 🚀

philphilphil commented:
Glad you like it :).

It's not in line which was on purpose to distinguish it from the filter options. But actually I think now it might be better to keep it consistent. I would change it to use the sentence-syntax and add everything to README.

With the requirements from #81 I would implement the following rules:
- hide task count
- hide backlinks
- hide done date
- hide due date
- hide recurrence rule
- hide edit button

schemar commented:
That sounds awesome. Looking very much forward to it! Let me know if you want to discuss possible options of how to make those optional. My initial guess would be an optional argument to the Tasks toLi method.

philphilphil commented:
It's implemented. I'm fairly new to TS, so please let me know if its good like that of if it needs some changes.

I implemented a new class "LayoutOption" which is filled from Query and used in QueryRenderer where it is passed into the task.toLi.
There were issues with the "due date" and "done date" filters. Those two did also match on "hide due date" and "hide done date" I fixed this by removing the ? from both regexp. I tested a lot, and it doesn't seem to affect everything else, but please check if that's okay.

schemar commented:
Thanks a bunch! Though it may take a while until I have the time to review your changes.

schemar commented on src/Task.ts:
```diff
@@ -263,18 +266,32 @@ export class Task {
+
+        var taskString = this.description;
+
+        if (layoutOptions == undefined || !layoutOptions.hideRecurrenceRule) {
```
Please use triple `===` for comparison. JavaScript is generally horrible when it comes to `==` …

schemar replied:
To simplify all the `if`s, you could add something like `layoutOptions = layoutOptions ?? new LayoutOptions()` at the beginning of the function. That way, `layoutOptions` will always be defined.

(`??` is a typescript operator; it relates to JavaScript’s horrible way of handling `null` and `undefined` …)

schemar commented on src/Task.ts:
```diff
@@ -263,18 +266,32 @@ export class Task {
-        return `${this.description}${recurrenceRule}${dueDate}${doneDate}${this.blockLink}`;
+    public toString(layoutOptions?: LayoutOptions): string {
+
+        var taskString = this.description;
```
Please always use `let` and `cons` everywhere instead of `var` (another JS quirk).

schemar commented on src/Task.ts:
```diff
@@ -263,18 +266,32 @@ export class Task {
-            : '';
-
-        return `${this.description}${recurrenceRule}${dueDate}${doneDate}${this.blockLink}`;
+    public toString(layoutOptions?: LayoutOptions): string {
```
Personally I would prefer `toString({layoutOptions}: {layoutOptions?: LayoutOptions})`, but it’s ok if you leave it as is. It’s especially fine the way it is when there is only a single argument.

philphilphil replied:
I left it as it is since I did not find out how to make it optional like this.

schemar reviewed (changes requested):
Thank you so much for your time. Here comes my review.

Overall it is very good. Just some minor things that have to be done in order to preserve some existing functionality: the Query regexes need to be fixed in a different way. The correct fix is starting the due and done regexes with `^` instead of removing the `?`. This way it won’t match accidentally, yet queries like `due today` will still work. Actually, *all* regexes should probably start with `/^` …

In the `QueryRenderer`, I think it would be more readable if you extracted the content inside `if (!this.query.layoutOptions.hideBacklinks) {…}` and `if (!this.query.layoutOptions.hideEditButton) ` into their own functions. If it’s feasible. I am not sure which dependencies there are. What do you think?

And some minor JS-specific remarks in-line.

philphilphil commented:
Thanks for the review!
- Regex are fixed (and ^ added to all of them). 
- QueryRenderer stuff refactored into their own methods.
- Stuff from conversations fixed.

schemar commented:
Thank you! I hope this time I will get to it sooner.

schemar commented on src/Task.ts:
```diff
@@ -4,6 +4,7 @@ import { RRule } from 'rrule';
 import { replaceTaskWithTasks } from './File';
 import { getSettings } from './Settings';
 import type { Moment } from 'moment';
+import { LayoutOptions } from 'LayoutOptions';
```
Please import with relative path (jest fails at the moment).
```suggestion
import { LayoutOptions } from './LayoutOptions';
```

schemar reviewed (approved):
Thank you! :heart:
